# Guided Capstone Step 4. Pre-Processing and Training Data Development

**The Data Science Method**  


1.   Problem Identification 


2.   Data Wrangling 
  
 
3.   Exploratory Data Analysis   

4.   **Pre-processing and Training Data Development**  
 * Create dummy or indicator features for categorical variables
  * Standardize the magnitude of numeric features
  * Split into testing and training datasets
  * Apply scaler to the testing set
5.   Modeling 
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

**<font color='teal'> Start by loading the necessary packages as we did in step 3 and printing out our current working directory just to confirm we are in the correct project directory. </font>**

In [2]:
import os
import pandas as pd
import datetime
import seaborn as sns
import matplotlib.pyplot as plt
import numpy as np
import sklearn.metrics as skm
%matplotlib inline

**<font color='teal'>  Load the csv file you created in step 3, remember it should be saved inside your data subfolder and print the first five rows.</font>**

In [3]:
path="/Users/rachelruzzo/Documents/GitHub/Guided-Capstone/Data Wrangling/data"
os.listdir(path)
df = pd.read_csv("/Users/rachelruzzo/Documents/GitHub/Guided-Capstone/EDA/step3_output.csv")
sd = df.drop(columns=["Unnamed: 0"])
sd= sd.drop(166, axis=0)
sd.tail(5)

Name    state  summit_elev  vertical_drop  \
161                      Hogadon Basin  Wyoming         8000            640   
162          Sleeping Giant Ski Resort  Wyoming         7428            810   
163                   Snow King Resort  Wyoming         7808           1571   
164  Snowy Range Ski & Recreation Area  Wyoming         9663            990   
165                White Pine Ski Area  Wyoming         9500           1100   

     fastQuads  quad  triple  double  surface  total_chairs  ...  \
161          0     0       0       1        1             2  ...   
162          0     0       1       1        1             3  ...   
163          0     1       1       1        0             3  ...   
164          0     0       1       3        1             5  ...   
165          0     0       2       0        0             2  ...   

     SkiableTerrain_ac  Snow Making_ac  daysOpenLastYear  yearsOpen  \
161               92.0            32.0        121.000000       61.0   
162              184.0            18.0         61.000000       81.0   
163              400.0           250.0        121.000000       80.0   
164               75.0            30.0        131.000000       59.0   
165              370.0             0.0        115.103943       81.0   

     averageSnowfall  AdultWeekday  AdultWeekend  projectedDaysOpen  \
161             80.0     48.000000          48.0          95.000000   
162            310.0     42.000000          42.0          77.000000   
163            300.0     59.000000          59.0         123.000000   
164            250.0     49.000000          49.0         120.053004   
165            150.0     57.916957          49.0         120.053004   

     NightSkiing_ac  clusters  
161             0.0       1.0  
162             0.0       1.0  
163           110.0       1.0  
164             0.0       1.0  
165             0.0       1.0  

[5 rows x 23 columns]

## Create dummy features for categorical variables

**<font color='teal'> Create dummy variables for `state`. Add the dummies back to the dataframe and remove the original column for `state`. </font>**

Hint: you can see an example of how to execute this in Aiden's article on preprocessing [here](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-pre-processing-and-training-data-development-fd2d75182967). 

In [4]:
sdo = sd.state
sd = pd.concat([sd.drop(columns="state"), pd.get_dummies(sdo)], axis=1)
sd.head()

Name  summit_elev  vertical_drop  fastQuads  quad  \
0               Hilltop Ski Area         2090            294          0     0   
1            Sunrise Park Resort        11100           1800          1     2   
2  Yosemite Ski & Snowboard Area         7800            600          0     0   
3               Donner Ski Ranch         8012            750          0     0   
4                  June Mountain        10090           2590          2     0   

   triple  double  surface  total_chairs  Runs  ...  Rhode Island  \
0       1       0        2             3  13.0  ...             0   
1       3       1        0             7  65.0  ...             0   
2       1       3        1             5  10.0  ...             0   
3       1       5        2             8  52.0  ...             0   
4       0       4        1             7  41.0  ...             0   

   South Dakota  Tennessee  Utah  Vermont  Virginia  Washington  \
0             0          0     0        0         0           0   
1             0          0     0        0         0           0   
2             0          0     0        0         0           0   
3             0          0     0        0         0           0   
4             0          0     0        0         0           0   

   West Virginia  Wisconsin  Wyoming  
0              0          0        0  
1              0          0        0  
2              0          0        0  
3              0          0        0  
4              0          0        0  

[5 rows x 57 columns]

## Standardize the magnitude of numeric features

**<font color='teal'> Using sklearn preprocessing standardize the scale of the features of the dataframe except the name of the resort which we done't need in the dataframe for modeling, so it can be droppped here as well. Also, we want to hold out our response variable(s) so we can have their true values available for model performance review. Let's set `AdultWeekend` to the y variable as our response for scaling and modeling. Later we will go back and consider the `AdultWeekday`, `dayOpenLastYear`, and `projectedDaysOpen`. For now leave them in the development dataframe. </font>**

In [5]:
# first we import the preprocessing package from the sklearn library
from sklearn import preprocessing

# Declare an explanatory variable, called X,and assign it the result of dropping 'Name' and 'AdultWeekend' from the df
X= sd.drop(['Name','AdultWeekend'], axis=1)

# Declare a response variable, called y, and assign it the AdultWeekend column of the df 
y = sd.AdultWeekend 

# Here we use the StandardScaler() method of the preprocessing package, and then call the fit() method with parameter X 
scaler = preprocessing.StandardScaler().fit(X)

# Declare a variable called X_scaled, and assign it the result of calling the transform() method with parameter X 
X_scaled=scaler.transform(X) 

In [13]:
xscaledf = pd.DataFrame(X_scaled)
xscaledf

0         1         2         3         4         5         6   \
0   -0.472894 -0.937833 -0.395389 -0.765798 -0.113312 -1.207291 -0.149174   
1    2.315714  1.652138  1.378517  2.059146  1.596666 -0.473187 -1.564197   
2    1.294359 -0.411584 -0.395389 -0.765798 -0.113312  0.995020 -0.856685   
3    1.359973 -0.153619 -0.395389 -0.765798 -0.113312  2.463228 -0.149174   
4    2.003117  3.010755  3.152422 -0.765798 -0.968301  1.729124 -0.856685   
..        ...       ...       ...       ...       ...       ...       ...   
161  1.356259 -0.342793 -0.395389 -0.765798 -0.968301 -0.473187 -0.856685   
162  1.179225 -0.050433 -0.395389 -0.765798 -0.113312 -0.473187 -0.856685   
163  1.296835  1.258311 -0.395389  0.646674 -0.113312 -0.473187 -1.564197   
164  1.870960  0.259126 -0.395389 -0.765798 -0.113312  0.995020 -0.856685   
165  1.820511  0.448300 -0.395389 -0.765798  0.741677 -1.207291 -1.564197   

           7         8         9   ...       45       46       47        48  \
0   -1.110136 -0.957818 -1.188728  ... -0.07785 -0.07785 -0.07785 -0.135665   
1    0.502841  2.078496 -0.124642  ... -0.07785 -0.07785 -0.07785 -0.135665   
2   -0.303648 -1.132990 -0.124642  ... -0.07785 -0.07785 -0.07785 -0.135665   
3    0.906085  1.319418 -0.124642  ... -0.07785 -0.07785 -0.07785 -0.135665   
4    0.502841  0.677120 -0.124642  ... -0.07785 -0.07785 -0.07785 -0.135665   
..        ...       ...       ...  ...      ...      ...      ...       ...   
161 -1.513380 -0.081958 -1.188728  ... -0.07785 -0.07785 -0.07785 -0.135665   
162 -1.110136  1.085855 -1.188728  ... -0.07785 -0.07785 -0.07785 -0.135665   
163 -1.110136  0.151605 -0.124642  ... -0.07785 -0.07785 -0.07785 -0.135665   
164 -0.303648  0.209995 -0.124642  ... -0.07785 -0.07785 -0.07785 -0.135665   
165 -1.513380 -0.257130  0.748748  ... -0.07785 -0.07785 -0.07785 -0.135665   

           49        50        51        52        53        54  
0   -0.209822 -0.110432 -0.157135 -0.135665 -0.239426 -0.176227  
1   -0.209822 -0.110432 -0.157135 -0.135665 -0.239426 -0.176227  
2   -0.209822 -0.110432 -0.157135 -0.135665 -0.239426 -0.176227  
3   -0.209822 -0.110432 -0.157135 -0.135665 -0.239426 -0.176227  
4   -0.209822 -0.110432 -0.157135 -0.135665 -0.239426 -0.176227  
..        ...       ...       ...       ...       ...       ...  
161 -0.209822 -0.110432 -0.157135 -0.135665 -0.239426  5.674504  
162 -0.209822 -0.110432 -0.157135 -0.135665 -0.239426  5.674504  
163 -0.209822 -0.110432 -0.157135 -0.135665 -0.239426  5.674504  
164 -0.209822 -0.110432 -0.157135 -0.135665 -0.239426  5.674504  
165 -0.209822 -0.110432 -0.157135 -0.135665 -0.239426  5.674504  

[166 rows x 55 columns]

## Split into training and testing datasets

**<font color='teal'> Using sklearn model selection import train_test_split, and create a 75/25 split with the y = `AdultWeekend`. We will start by using the adult weekend ticket price as our response variable for modeling.</font>**

In [5]:
# Import the train_test_split function from the sklearn.model_selection utility.  
from sklearn.model_selection import train_test_split

# Get the 1-dimensional flattened array of our response variable y by calling the ravel() function on y
y = y.ravel()

# Call the train_test_split() function with the first two parameters set to X_scaled and y 
# Declare four variables, X_train, X_test, y_train and y_test separated by commas 
X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)

Here we start the actual modeling work. First let's fit a multiple linear regression model to predict the `AdultWeekend` price.

# Guided Capstone Step 5. Modeling


This is the fifth step in the Data Science Method. In the previous steps you cleaned and prepared the datasets. Now it's time to get into the most exciting part: modeling! In this exercise, you'll build three different models and compare each model's performance. In the end, you'll choose the best model for demonstrating insights to Big Mountain management.



### **The Data Science Method**  


1.   Problem Identification 

2.   Data Wrangling 
  
3.   Exploratory Data Analysis 
 
4.   Pre-processing and Training Data Development

5.   **Modeling**
  * Fit Models with Training Data Set
  * Review Model Outcomes — Iterate over additional models as needed.
  * Identify the Final Model

6.   Documentation
  * Review the Results
  * Present and share your findings - storytelling
  * Finalize Code 
  * Finalize Documentation

## Fit Models with a Training Dataset

**<font color='teal'> Using sklearn, fit the model on your training dataset.</font>**

#### Model 1

In [6]:
#all first model set
from sklearn import linear_model
from sklearn.metrics import explained_variance_score,mean_absolute_error
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

**<font color='teal'> Predict on the testing dataset and score the model performance with the y_test set and the y-pred values. The explained variance is a measure of the variation explained by the model. This is also known as the R-squared value. </font>**

Hint: you will have to use the `predict()` method here as it's used in this [DSM article](https://medium.com/@aiden.dataminer/the-data-science-method-dsm-modeling-56b4233cad1b) about modeling. 

In [7]:
# Make a variable called y_pred and assign it the result of calling predict() on our model variable with parameter X_test
y_pred = model.predict(X_test)
y_pred

array([5.15258789e+01, 5.10297852e+01, 7.84711914e+01, 6.45903320e+01,
       6.92309570e+01, 6.91528320e+01, 5.55415039e+01, 4.04340820e+01,
       4.04516602e+01, 6.79184570e+01, 5.52602539e+01, 4.56000977e+01,
       3.59262695e+01, 5.08403320e+01, 4.57348633e+01, 6.13208008e+01,
       4.03598633e+01, 5.49145508e+01, 7.05786133e+01, 4.61272228e+13,
       5.49086914e+01, 4.37700195e+01, 4.61282481e+13, 4.61273169e+13,
       6.36450195e+01, 5.88247070e+01, 6.21723633e+01, 5.08950195e+01,
       6.87133789e+01, 6.70415039e+01, 3.10434570e+01, 5.87602539e+01,
       6.70825195e+01, 7.29731445e+01, 6.14633789e+01, 4.00786133e+01,
       6.14497070e+01, 3.30805664e+01, 5.31215820e+01, 6.69653320e+01,
       6.00473633e+01, 4.04946289e+01])

## Review Model Outcomes — Iterate over additional models as needed

In [8]:
# You might want to use the explained_variance_score() and mean_absolute_error() metrics.
from sklearn.metrics import explained_variance_score, mean_absolute_error
# You can plug y_test and y_pred into the functions to evaluate the model
print(skm.mean_absolute_error(y_test, y_pred))
print(skm.explained_variance_score(y_test,y_pred))

3294828283263.3066
-8.579758406493125e+23


In [9]:
lm.coef_

array([ 1.37926035e+00,  9.21597045e-01,  7.41359526e+11,  9.31063888e+11,
        1.53815050e+12,  1.79143822e+12,  1.85877111e+12, -3.26130342e+12,
        1.51581319e+00,  5.72387695e-01, -8.17138672e-01, -5.60180664e-01,
        1.58984375e+00,  5.72387695e-01, -6.79931641e-01,  5.75622559e-01,
        9.61099243e+00, -5.82031250e-01,  5.90087891e-01, -2.23632812e-01,
       -3.56937418e+12, -3.56937418e+12, -9.87924243e+12, -7.88402596e+12,
       -7.88402596e+12, -8.60965447e+12, -6.14475449e+12, -5.03253759e+12,
       -6.14475449e+12, -7.88402596e+12,  7.20585418e+07, -7.07355295e+12,
       -1.32246507e+13, -9.87924243e+12, -5.03253759e+12, -6.14475449e+12,
       -3.56937418e+12, -1.14739327e+13, -7.27888241e+06, -8.60965447e+12,
       -1.46853930e+13, -6.14475449e+12, -7.07355295e+12, -6.14475449e+12,
       -1.09751990e+13,  6.85198205e+02, -3.56937418e+12, -3.56937418e+12,
       -6.14475449e+12, -9.27038323e+12, -5.03253759e+12, -7.07355295e+12,
       -6.14475449e+12, -

**<font color='teal'> Print the intercept value from the linear model. </font>**

Hint: our linear regression model `lm` has an attribute `intercept_` for the intercept

In [10]:
lm.intercept_

833631252447.1724

**<font color='teal'> The intercept is the mean `AdultWeekend` price for all the resorts given the other characteristics. The addition or subtraction of each of the coefficient values in the regression are numeric adjustments applied to the intercept to provide a particular observation's value for the resulting `AdultWeekend` value. Also, because we took the time to scale our x values in the training data, we can compare each of the coeeficients for the features to determine the feature importances. Print the coefficient values from the linear model and sort in descending order to identify the top ten most important features.</font>** 


Hint: make sure to review the absolute value of the coefficients, because the adjustment may be positive or negative, but what we are looking for is the magnitude of impact on our response variable.

In [11]:
# You might want to make a pandas DataFrame displaying the coefficients for each state like so: 
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending = False)

Coefficient
New York           1.468539e+13
Michigan           1.322465e+13
New Hampshire      1.147393e+13
Pennsylvania       1.097520e+13
Wisconsin          1.044531e+13
California         9.879242e+12
Minnesota          9.879242e+12
Vermont            9.270383e+12
New Mexico         8.609654e+12
Idaho              8.609654e+12
Wyoming            7.884026e+12
Colorado           7.884026e+12
Maine              7.884026e+12
Connecticut        7.884026e+12
Ohio               7.073553e+12
Washington         7.073553e+12
Massachusetts      7.073553e+12
Montana            6.144754e+12
Iowa               6.144754e+12
Utah               6.144754e+12
Oregon             6.144754e+12
Illinois           6.144754e+12
West Virginia      6.144754e+12
North Carolina     6.144754e+12
Indiana            5.032538e+12
Missouri           5.032538e+12
Virginia           5.032538e+12
South Dakota       3.569374e+12
Arizona            3.569374e+12
Alaska             3.569374e+12
Nevada             3.569374e+12
Tennessee          3.569374e+12
total_chairs       3.261303e+12
surface            1.858771e+12
double             1.791438e+12
triple             1.538151e+12
quad               9.310639e+11
fastQuads          7.413595e+11
Maryland           7.205854e+07
New Jersey         7.278882e+06
Rhode Island       6.851982e+02
AdultWeekday       9.610992e+00
Snow Making_ac     1.589844e+00
Runs               1.515813e+00
summit_elev        1.379260e+00
vertical_drop      9.215970e-01
LongestRun_mi      8.171387e-01
yearsOpen          6.799316e-01
NightSkiing_ac     5.900879e-01
projectedDaysOpen  5.820312e-01
averageSnowfall    5.756226e-01
TerrainParks       5.723877e-01
daysOpenLastYear   5.723877e-01
SkiableTerrain_ac  5.601807e-01
clusters           2.236328e-01

**<font color='teal'>You should see that the top ten important features are different states. However, the state is not something the managers at the Big Mountain Resort can do anything about. Given that we care more about actionable traits associated with ticket pricing, rebuild the model without the state features and compare the results. </font>**

Hint: Try to construct another model using exactly the steps we followed above. 

#### Model 2

In [12]:
sd2 = pd.read_csv("/Users/rachelruzzo/Documents/GitHub/Guided-Capstone/EDA/step3_output.csv")
sd2 = sd2.drop(columns=["Unnamed: 0", "state"])

X= sd2.drop(['Name','AdultWeekend'], axis=1)
y = sd2.AdultWeekend 

scaler = preprocessing.StandardScaler().fit(X)

X_scaled=scaler.transform(X) 

y = y.ravel()

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [13]:
y_pred = model.predict(X_test)
y_pred
print(skm.mean_absolute_error(y_test, y_pred))
print(skm.explained_variance_score(y_test,y_pred))
print(lm.intercept_)
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending = False)

6.868294649676805
0.5587146561151805
56.454375357298126


Coefficient
AdultWeekday         10.919201
clusters              2.599700
summit_elev           2.535971
vertical_drop         2.208263
daysOpenLastYear      1.866794
averageSnowfall       1.801860
SkiableTerrain_ac     1.364137
triple                1.296569
Snow Making_ac        1.267853
surface               1.175355
double                0.963353
Runs                  0.690440
projectedDaysOpen     0.576397
total_chairs          0.546326
quad                  0.386995
LongestRun_mi         0.386497
NightSkiing_ac        0.346109
fastQuads             0.210240
yearsOpen             0.153519
TerrainParks          0.147613


**<font color='teal'> When reviewing our new model coefficients, we see `summit_elev` is now in the number two spot. This is also difficult to change from a management prespective and highly correlated with `base_elev` and `vertical_drop`.  This time, rebuild the model without the state features and without the `summit_elev` and without `base_elev`and compare the results. </font>**

#### Model 3

In [14]:
sd3 = pd.read_csv("/Users/rachelruzzo/Documents/GitHub/Guided-Capstone/EDA/step3_output.csv")
sd3 = sd3.drop(columns=["Unnamed: 0", "state", "summit_elev"])

X= sd3.drop(['Name','AdultWeekend'], axis=1)
y = sd3.AdultWeekend 

scaler = preprocessing.StandardScaler().fit(X)

X_scaled=scaler.transform(X) 

y = y.ravel()

X_train, X_test, y_train, y_test = train_test_split(X_scaled, y, test_size=0.25, random_state=1)
lm = linear_model.LinearRegression()
model = lm.fit(X_train,y_train)

In [15]:
y_pred = model.predict(X_test)
y_pred
print(skm.mean_absolute_error(y_test, y_pred))
print(skm.explained_variance_score(y_test,y_pred))
print(lm.intercept_)
pd.DataFrame(abs(lm.coef_), X.columns, columns=['Coefficient']).sort_values('Coefficient', ascending = False)

7.331908540674727
0.5213130038510798
56.41196935322093


Coefficient
AdultWeekday         10.510855
averageSnowfall       2.403102
SkiableTerrain_ac     2.369025
vertical_drop         2.363385
clusters              2.307149
Snow Making_ac        1.973244
daysOpenLastYear      1.468361
triple                1.229835
double                1.035501
surface               0.889893
total_chairs          0.654035
TerrainParks          0.378878
LongestRun_mi         0.362764
yearsOpen             0.283120
projectedDaysOpen     0.278217
fastQuads             0.240221
quad                  0.210317
Runs                  0.189953
NightSkiing_ac        0.075843

## Identify the Final Model

**<font color='teal'> Review the model performances in the table below and choose the best model for proving insights to Big Mountain management about what features are driving ski resort lift ticket prices. Type your choice in the final markdown cell — you will discuss this selection more in the next step of the guided casptone. </font>**

| Model | Explained Variance| Mean Absolute Error|Features Dropped|
| --- | --- | --- | --- |
| Model 1. | -8.579758406493125e+23 | 3294828283263.3066 |-|
| Model 2. | 0.5587146561151805 | 6.868294649676805 |'state'|
| Model 3. | 0.5213130038510798 | 7.331908540674727|'state','summit_elev'|

Model Selection:

Model 2